telegram/ext/filters.py

```diff
@@ -865,73 +865,83 @@ class user(BaseFilter):
             user_ids(set(:obj:`int`), optional): Which user ID(s) to allow through.
             usernames(set(:obj:`str`), optional): Which username(s) (without leading '@') to allow
                 through.
+            allow_empty(:obj:`bool`, optional): Whether updates should be processed, if no user
+                is specified in :attr:`user_ids` and :attr:`usernames`.
         Args:
             user_id(:obj:`int` | List[:obj:`int`], optional): Which user ID(s) to allow
                 through.
             username(:obj:`str` | List[:obj:`str`], optional): Which username(s) to allow
                 through. If username starts with '@' symbol, it will be ignored.
+            allow_empty(:obj:`bool`, optional): Whether updates should be processed, if no user
+                is specified in :attr:`user_ids` and :attr:`usernames`. Defaults to :obj:`False`
 
         Raises:
             ValueError: If chat_id and username are both present, or neither is.
 
         """
 
-        def __init__(self, user_id=None, username=None):
-            if (user_id is None) == (username is None):
-                raise ValueError('One and only one of user_id or username must be used')
+        def __init__(self, user_id=None, username=None, allow_empty=False):
+            self.allow_empty = allow_empty
+            self.__lock = Lock()
 
-            self._user_ids_lock = Lock()
-            self._usernames_lock = Lock()
+            self._user_ids = {}
+            self._usernames = {}
 
-            # Initialize in a way that will not fail the first setter calls
-            self._user_ids = user_id
-            self._usernames = username
-            # Actually initialize
             self.user_ids = user_id
             self.usernames = username
 
+        def _set_user_ids(self, user_id):
+            if user_id is None:
+                self._user_ids = {}
+            elif isinstance(user_id, int):
+                self._user_ids = {user_id}
+            else:
+                self._user_ids = set(user_id)
+
+        def _set_usernames(self, username):
+            if username is None:
+                self._usernames = {}
+            elif isinstance(username, str):
+                self._usernames = {username.replace('@', '')}
+            else:
+                self._usernames = set([user.replace('@', '') for user in username])
+
         @property
         def user_ids(self):
-            with self._user_ids_lock:
-                return self._user_ids
+            with self.__lock:
+                return frozenset(self._user_ids) if self._user_ids is not None else None
 
         @user_ids.setter
         def user_ids(self, user_id):
```

Comment:
Documentation needs to clearly specify that this will completely replace the set of user_ids.

Thinking of it, we might need to add a proper function to add to user_ids (and user_names) and another to remove:

`def add_user_id(self, user_id):`
`def remove_user_id(self, user_id):`

Rationale:
If a user "polls" the current user_ids and creates a new Set to set, by this time, a different thread might have updated the user_ids... (race condition).
So we need to provide granular operations which will be useful to users.

Reply:
Sounds good. will do.

Reply:
Left some comments on that code